In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import scipy.stats as ss

In [ ]:
df = pd.read_csv("D:\Python\Python_Journey\Data Analysis\Statistics\Exploratory Data Analysis\EDA Practice\Students Performance\StudentPerformanceFactors.csv")

<>:1: SyntaxWarning: "\P" is an invalid escape sequence. Such sequences will not work in the future. Did you mean "\\P"? A raw string is also an option.
<>:1: SyntaxWarning: "\P" is an invalid escape sequence. Such sequences will not work in the future. Did you mean "\\P"? A raw string is also an option.
C:\Users\Admin\AppData\Local\Temp\ipykernel_8404\2848596265.py:1: SyntaxWarning: "\P" is an invalid escape sequence. Such sequences will not work in the future. Did you mean "\\P"? A raw string is also an option.
  df = pd.read_csv("D:\Python\Python_Journey\Data Analysis\Statistics\Exploratory Data Analysis\EDA Practice\Students Performance\StudentPerformanceFactors.csv")


,Hours_Studied,Attendance,Parental_Involvement,Access_to_Resources,Extracurricular_Activities,Sleep_Hours,Previous_Scores,Motivation_Level,Internet_Access,Tutoring_Sessions,Family_Income,Teacher_Quality,School_Type,Peer_Influence,Physical_Activity,Learning_Disabilities,Parental_Education_Level,Distance_from_Home,Gender,Exam_Score
0,23.0,84.0,Low,High,No,7,73,Low,Yes,0.0,Low,Medium,Public,Positive,3.0,No,High School,Near,Male,67.0
1,19.0,64.0,Low,Medium,No,8,59,Low,Yes,NaN,Medium,Medium,Public,Negative,4.0,No,College,Moderate,Female,61.0
2,24.0,98.0,Medium,Medium,Yes,7,91,Medium,Yes,2.0,Medium,Medium,Public,Neutral,4.0,No,Postgraduate,Near,Male,74.0
3,29.0,89.0,Low,Medium,Yes,8,98,Medium,Yes,1.0,Medium,Medium,Public,Negative,4.0,No,High School,Moderate,Male,71.0
4,19.0,92.0,Medium,Medium,Yes,6,65,Medium,Yes,3.0,Medium,High,Public,Neutral,4.0,No,College,Near,Female,70.0


In [7]:
df.head()

,Hours_Studied,Attendance,Parental_Involvement,Access_to_Resources,Extracurricular_Activities,Sleep_Hours,Previous_Scores,Motivation_Level,Internet_Access,Tutoring_Sessions,Family_Income,Teacher_Quality,School_Type,Peer_Influence,Physical_Activity,Learning_Disabilities,Parental_Education_Level,Distance_from_Home,Gender,Exam_Score
0,23.0,84.0,Low,High,No,7,73,Low,Yes,0.0,Low,Medium,Public,Positive,3.0,No,High School,Near,Male,67.0
1,19.0,64.0,Low,Medium,No,8,59,Low,Yes,NaN,Medium,Medium,Public,Negative,4.0,No,College,Moderate,Female,61.0
2,24.0,98.0,Medium,Medium,Yes,7,91,Medium,Yes,2.0,Medium,Medium,Public,Neutral,4.0,No,Postgraduate,Near,Male,74.0
3,29.0,89.0,Low,Medium,Yes,8,98,Medium,Yes,1.0,Medium,Medium,Public,Negative,4.0,No,High School,Moderate,Male,71.0
4,19.0,92.0,Medium,Medium,Yes,6,65,Medium,Yes,3.0,Medium,High,Public,Neutral,4.0,No,College,Near,Female,70.0


In [9]:
df.describe()

,Hours_Studied,Attendance,Sleep_Hours,Previous_Scores,Tutoring_Sessions,Physical_Activity,Exam_Score
count,6605.000000,6605.000000,6607.00000,6607.000000,6606.000000,6604.000000,6605.000000
mean,19.973808,79.948978,7.02906,75.070531,1.493642,2.967293,67.215291
std,5.995209,11.736506,1.46812,14.399784,1.230648,1.031015,4.217186
min,0.000000,-91.000000,4.00000,50.000000,0.000000,0.000000,-65.000000
25%,16.000000,70.000000,6.00000,63.000000,1.000000,2.000000,65.000000
50%,20.000000,80.000000,7.00000,75.000000,1.000000,3.000000,67.000000
75%,24.000000,90.000000,8.00000,88.000000,2.000000,4.000000,69.000000
max,44.000000,100.000000,10.00000,100.000000,8.000000,6.000000,101.000000


In [10]:
df.corr(numeric_only=True)

,Hours_Studied,Attendance,Sleep_Hours,Previous_Scores,Tutoring_Sessions,Physical_Activity,Exam_Score
Hours_Studied,1.000000,-0.011679,0.011353,0.025065,-0.014760,0.004203,0.444881
Attendance,-0.011679,1.000000,-0.015616,-0.020886,0.015156,-0.026294,0.522748
Sleep_Hours,0.011353,-0.015616,1.000000,-0.021750,-0.012258,-0.000882,-0.018976
Previous_Scores,0.025065,-0.020886,-0.021750,1.000000,-0.013054,-0.010514,0.156261
Tutoring_Sessions,-0.014760,0.015156,-0.012258,-0.013054,1.000000,0.017438,0.142553
Physical_Activity,0.004203,-0.026294,-0.000882,-0.010514,0.017438,1.000000,0.020787
Exam_Score,0.444881,0.522748,-0.018976,0.156261,0.142553,0.020787,1.000000


In [11]:
# Notable observations
# Exam Score is 44% related to hours studied
# Exam Score is 52% related to Attendence studied

In [12]:
df.head()

,Hours_Studied,Attendance,Parental_Involvement,Access_to_Resources,Extracurricular_Activities,Sleep_Hours,Previous_Scores,Motivation_Level,Internet_Access,Tutoring_Sessions,Family_Income,Teacher_Quality,School_Type,Peer_Influence,Physical_Activity,Learning_Disabilities,Parental_Education_Level,Distance_from_Home,Gender,Exam_Score
0,23.0,84.0,Low,High,No,7,73,Low,Yes,0.0,Low,Medium,Public,Positive,3.0,No,High School,Near,Male,67.0
1,19.0,64.0,Low,Medium,No,8,59,Low,Yes,NaN,Medium,Medium,Public,Negative,4.0,No,College,Moderate,Female,61.0
2,24.0,98.0,Medium,Medium,Yes,7,91,Medium,Yes,2.0,Medium,Medium,Public,Neutral,4.0,No,Postgraduate,Near,Male,74.0
3,29.0,89.0,Low,Medium,Yes,8,98,Medium,Yes,1.0,Medium,Medium,Public,Negative,4.0,No,High School,Moderate,Male,71.0
4,19.0,92.0,Medium,Medium,Yes,6,65,Medium,Yes,3.0,Medium,High,Public,Neutral,4.0,No,College,Near,Female,70.0


In [ ]:
# pd.crosstab(df["Parental_Involvement"], df["Access_to_Resources"])
# df.value_counts(df["Parental_Involvement"])
# df.value_counts(df["Access_to_Resources"])
df.value_counts(df["Gender"])

# Maximum parents are moderately involved
# Maximum Studenst in this sample have moderate access to resources
# Maximum students in this sample have moderate motivation to study
# Maximum students in this sample have Access to internet
# Low and Medium income students tie each other in number
# Max are from public school
# Max parents are high school while college and pg are tied
# Max are near
# Gender are most male but both are close

Gender
Male      3813
Female    2793
Name: count, dtype: int64

In [33]:
df.head()

,Hours_Studied,Attendance,Parental_Involvement,Access_to_Resources,Extracurricular_Activities,Sleep_Hours,Previous_Scores,Motivation_Level,Internet_Access,Tutoring_Sessions,Family_Income,Teacher_Quality,School_Type,Peer_Influence,Physical_Activity,Learning_Disabilities,Parental_Education_Level,Distance_from_Home,Gender,Exam_Score
0,23.0,84.0,Low,High,No,7,73,Low,Yes,0.0,Low,Medium,Public,Positive,3.0,No,High School,Near,Male,67.0
1,19.0,64.0,Low,Medium,No,8,59,Low,Yes,NaN,Medium,Medium,Public,Negative,4.0,No,College,Moderate,Female,61.0
2,24.0,98.0,Medium,Medium,Yes,7,91,Medium,Yes,2.0,Medium,Medium,Public,Neutral,4.0,No,Postgraduate,Near,Male,74.0
3,29.0,89.0,Low,Medium,Yes,8,98,Medium,Yes,1.0,Medium,Medium,Public,Negative,4.0,No,High School,Moderate,Male,71.0
4,19.0,92.0,Medium,Medium,Yes,6,65,Medium,Yes,3.0,Medium,High,Public,Neutral,4.0,No,College,Near,Female,70.0


In [46]:
df.pivot_table(
    index = "Motivation_Level",
    values = "Sleep_Hours",
    aggfunc= ["mean", "median"],
    fill_value= 0
)

# Parental involvment is not related to anything numerically

,mean,median
,Sleep_Hours,Sleep_Hours
Motivation_Level,,
High,7.030326,7.0
Low,7.027376,7.0
Medium,7.030140,7.0


In [47]:
df.head()

,Hours_Studied,Attendance,Parental_Involvement,Access_to_Resources,Extracurricular_Activities,Sleep_Hours,Previous_Scores,Motivation_Level,Internet_Access,Tutoring_Sessions,Family_Income,Teacher_Quality,School_Type,Peer_Influence,Physical_Activity,Learning_Disabilities,Parental_Education_Level,Distance_from_Home,Gender,Exam_Score
0,23.0,84.0,Low,High,No,7,73,Low,Yes,0.0,Low,Medium,Public,Positive,3.0,No,High School,Near,Male,67.0
1,19.0,64.0,Low,Medium,No,8,59,Low,Yes,NaN,Medium,Medium,Public,Negative,4.0,No,College,Moderate,Female,61.0
2,24.0,98.0,Medium,Medium,Yes,7,91,Medium,Yes,2.0,Medium,Medium,Public,Neutral,4.0,No,Postgraduate,Near,Male,74.0
3,29.0,89.0,Low,Medium,Yes,8,98,Medium,Yes,1.0,Medium,Medium,Public,Negative,4.0,No,High School,Moderate,Male,71.0
4,19.0,92.0,Medium,Medium,Yes,6,65,Medium,Yes,3.0,Medium,High,Public,Neutral,4.0,No,College,Near,Female,70.0


In [48]:
pd.crosstab(df["Parental_Involvement"], df["Extracurricular_Activities"])

Extracurricular_Activities,No,Yes
Parental_Involvement,,
High,808,1100
Low,526,808
Medium,1333,2029


In [49]:
df["Parental_Involvement"].value_counts()

Parental_Involvement
Medium    3362
High      1908
Low       1336
Name: count, dtype: int64

In [54]:
# Assuming your DataFrame 'df' has a column named 'Level' (or 'Category')
df2 = df.groupby("Parental_Involvement").sample(n=1000, replace=False, random_state=42)

In [56]:
df2["Parental_Involvement"].value_counts()
df2.head()

,Hours_Studied,Attendance,Parental_Involvement,Access_to_Resources,Extracurricular_Activities,Sleep_Hours,Previous_Scores,Motivation_Level,Internet_Access,Tutoring_Sessions,Family_Income,Teacher_Quality,School_Type,Peer_Influence,Physical_Activity,Learning_Disabilities,Parental_Education_Level,Distance_from_Home,Gender,Exam_Score
4915,21.0,85.0,High,High,Yes,6,74,Medium,Yes,2.0,Medium,Medium,Public,Neutral,4.0,No,High School,Near,Female,71.0
449,14.0,62.0,High,High,Yes,8,65,Low,Yes,3.0,Low,Medium,Public,Negative,2.0,No,High School,Far,Male,61.0
6361,27.0,68.0,High,Medium,No,8,70,High,Yes,0.0,Medium,Medium,Private,Positive,3.0,No,Postgraduate,Near,Male,68.0
3352,23.0,72.0,High,Low,Yes,10,84,Low,Yes,2.0,Low,Medium,Public,Negative,2.0,No,High School,Moderate,Male,65.0
1553,15.0,86.0,High,Medium,Yes,8,77,Medium,Yes,0.0,Medium,Medium,Public,Positive,3.0,No,Postgraduate,Near,Male,69.0


In [62]:
pd.crosstab(df2["Parental_Involvement"], df2["Extracurricular_Activities"]) # no rel
pd.crosstab(df2["Parental_Involvement"], df2["Access_to_Resources"]) # no rel
pd.crosstab(df2["Parental_Involvement"], df2["Motivation_Level"]) # no major rel
pd.crosstab(df2["Parental_Involvement"], df2["Internet_Access"]) # no major rel
pd.crosstab(df2["Parental_Involvement"], df2["Family_Income"]) # no major rel







Family_Income,High,Low,Medium
Parental_Involvement,,,
High,185,423,392
Low,182,436,381
Medium,193,394,413


In [65]:
# 2. Automatically select all text/categorical columns
from itertools import combinations


cat_cols = df2.select_dtypes(include=['object', 'category']).columns

# 3. Loop through all pair combinations and calculate crosstabs
for col1, col2 in combinations(cat_cols, 2):
    print(f"\n=================== {col1} vs {col2} ===================")
    ct = pd.crosstab(df2[col1], df2[col2], margins=True, margins_name="Total")
    print(ct.to_markdown())

C:\Users\Admin\AppData\Local\Temp\ipykernel_8404\3350315537.py:5: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  cat_cols = df2.select_dtypes(include=['object', 'category']).columns



=================== Parental_Involvement vs Access_to_Resources ===================
| Parental_Involvement   |   High |   Low |   Medium |   Total |
|:-----------------------|-------:|------:|---------:|--------:|
| High                   |    301 |   217 |      482 |    1000 |
| Low                    |    308 |   154 |      538 |    1000 |
| Medium                 |    291 |   205 |      503 |     999 |
| Total                  |    900 |   576 |     1523 |    2999 |

=================== Parental_Involvement vs Extracurricular_Activities ===================
| Parental_Involvement   |   No |   Yes |   Total |
|:-----------------------|-----:|------:|--------:|
| High                   |  439 |   561 |    1000 |
| Low                    |  409 |   590 |     999 |
| Medium                 |  396 |   604 |    1000 |
| Total                  | 1244 |  1755 |    2999 |

=================== Parental_Involvement vs Motivation_Level ===================
| Parental_Involvement   |   High |   L

C:\Users\Admin\AppData\Local\Temp\ipykernel_8404\3350315537.py:5: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  cat_cols = df2.select_dtypes(include=['object', 'category']).columns

=================== Parental_Involvement vs Access_to_Resources ===================
| Parental_Involvement   |   High |   Low |   Medium |   Total |
|:-----------------------|-------:|------:|---------:|--------:|
| High                   |    301 |   217 |      482 |    1000 |
| Low                    |    308 |   154 |      538 |    1000 |
| Medium                 |    291 |   205 |      503 |     999 |
| Total                  |    900 |   576 |     1523 |    2999 |

=================== Parental_Involvement vs Extracurricular_Activities ===================
| Parental_Involvement   |   No |   Yes |   Total |
|:-----------------------|-----:|------:|--------:|
| High                   |  439 |   561 |    1000 |
| Low                    |  409 |   590 |     999 |
| Medium                 |  396 |   604 |    1000 |
| Total                  | 1244 |  1755 |    2999 |

=================== Parental_Involvement vs Motivation_Level ===================
| Parental_Involvement   |   High |   Low |   Medium |   Total |
|:-----------------------|-------:|------:|---------:|--------:|
| High                   |    179 |   300 |      521 |    1000 |
| Low                    |    207 |   272 |      521 |    1000 |
| Medium                 |    214 |   286 |      499 |     999 |
| Total                  |    600 |   858 |     1541 |    2999 |

=================== Parental_Involvement vs Internet_Access ===================
| Parental_Involvement   |   No |   Yes |   Total |
|:-----------------------|-----:|------:|--------:|
| High                   |   86 |   914 |    1000 |
| Low                    |   81 |   919 |    1000 |
| Medium                 |   74 |   925 |     999 |
| Total                  |  241 |  2758 |    2999 |

=================== Parental_Involvement vs Family_Income ===================
| Parental_Involvement   |   High |   Low |   Medium |   Total |
|:-----------------------|-------:|------:|---------:|--------:|
| High                   |    185 |   423 |      392 |    1000 |
| Low                    |    182 |   436 |      381 |     999 |
| Medium                 |    193 |   394 |      413 |    1000 |
| Total                  |    560 |  1253 |     1186 |    2999 |

=================== Parental_Involvement vs Teacher_Quality ===================
| Parental_Involvement   |   High |   Low |   Medium |   Total |
|:-----------------------|-------:|------:|---------:|--------:|
| High                   |    297 |   102 |      590 |     989 |
| Low                    |    287 |   110 |      589 |     986 |
| Medium                 |    303 |    86 |      597 |     986 |
| Total                  |    887 |   298 |     1776 |    2961 |

=================== Parental_Involvement vs School_Type ===================
| Parental_Involvement   |   Private |   Public |   Total |
|:-----------------------|----------:|---------:|--------:|
| High                   |       318 |      682 |    1000 |
| Low                    |       287 |      713 |    1000 |
| Medium                 |       314 |      685 |     999 |
| Total                  |       919 |     2080 |    2999 |

=================== Parental_Involvement vs Peer_Influence ===================
| Parental_Involvement   |   Negative |   Neutral |   Positive |   Total |
|:-----------------------|-----------:|----------:|-----------:|--------:|
| High                   |        201 |       376 |        423 |    1000 |
| Low                    |        237 |       380 |        383 |    1000 |
| Medium                 |        209 |       407 |        384 |    1000 |
| Total                  |        647 |      1163 |       1190 |    3000 |

=================== Parental_Involvement vs Learning_Disabilities ===================
| Parental_Involvement   |   No |   Yes |   Total |
|:-----------------------|-----:|------:|--------:|
| High                   |  884 |   116 |    1000 |
| Low                    |  894 |   106 |    1000 |
| Medium                 |  894 |   106 |    1000 |
| Total                  | 2672 |   328 |    3000 |

=================== Parental_Involvement vs Parental_Education_Level ===================
| Parental_Involvement   |   College |   High School |   Postgraduate |   Total |
|:-----------------------|----------:|--------------:|---------------:|--------:|
| High                   |       307 |           500 |            179 |     986 |
| Low                    |       305 |           470 |            210 |     985 |
| Medium                 |       273 |           511 |            196 |     980 |
| Total                  |       885 |          1481 |            585 |    2951 |

=================== Parental_Involvement vs Distance_from_Home ===================
| Parental_Involvement   |   Far |   Moderate |   Near |   Total |
|:-----------------------|------:|-----------:|-------:|--------:|
| High                   |    95 |        309 |    580 |     984 |
| Low                    |   113 |        314 |    563 |     990 |
| Medium                 |    98 |        294 |    604 |     996 |
| Total                  |   306 |        917 |   1747 |    2970 |

=================== Parental_Involvement vs Gender ===================
| Parental_Involvement   |   Female |   Male |   Total |
|:-----------------------|---------:|-------:|--------:|
| High                   |      453 |    547 |    1000 |
| Low                    |      410 |    590 |    1000 |
| Medium                 |      407 |    592 |     999 |
| Total                  |     1270 |   1729 |    2999 |

=================== Access_to_Resources vs Extracurricular_Activities ===================
| Access_to_Resources   |   No |   Yes |   Total |
|:----------------------|-----:|------:|--------:|
| High                  |  352 |   548 |     900 |
| Low                   |  245 |   331 |     576 |
| Medium                |  647 |   875 |    1522 |
| Total                 | 1244 |  1754 |    2998 |

=================== Access_to_Resources vs Motivation_Level ===================
| Access_to_Resources   |   High |   Low |   Medium |   Total |
|:----------------------|-------:|------:|---------:|--------:|
| High                  |    175 |   252 |      472 |     899 |
| Low                   |    122 |   169 |      285 |     576 |
| Medium                |    303 |   436 |      784 |    1523 |
| Total                 |    600 |   857 |     1541 |    2998 |

=================== Access_to_Resources vs Internet_Access ===================
| Access_to_Resources   |   No |   Yes |   Total |
|:----------------------|-----:|------:|--------:|
| High                  |   76 |   824 |     900 |
| Low                   |   42 |   534 |     576 |
| Medium                |  123 |  1399 |    1522 |
| Total                 |  241 |  2757 |    2998 |

=================== Access_to_Resources vs Family_Income ===================
| Access_to_Resources   |   High |   Low |   Medium |   Total |
|:----------------------|-------:|------:|---------:|--------:|
| High                  |    172 |   391 |      337 |     900 |
| Low                   |     99 |   241 |      235 |     575 |
| Medium                |    289 |   621 |      613 |    1523 |
| Total                 |    560 |  1253 |     1185 |    2998 |

=================== Access_to_Resources vs Teacher_Quality ===================
| Access_to_Resources   |   High |   Low |   Medium |   Total |
|:----------------------|-------:|------:|---------:|--------:|
| High                  |    259 |    85 |      544 |     888 |
| Low                   |    175 |    59 |      335 |     569 |
| Medium                |    453 |   154 |      896 |    1503 |
| Total                 |    887 |   298 |     1775 |    2960 |

=================== Access_to_Resources vs School_Type ===================
| Access_to_Resources   |   Private |   Public |   Total |
|:----------------------|----------:|---------:|--------:|
| High                  |       280 |      620 |     900 |
| Low                   |       176 |      400 |     576 |
| Medium                |       463 |     1059 |    1522 |
| Total                 |       919 |     2079 |    2998 |

=================== Access_to_Resources vs Peer_Influence ===================
| Access_to_Resources   |   Negative |   Neutral |   Positive |   Total |
|:----------------------|-----------:|----------:|-----------:|--------:|
| High                  |        215 |       328 |        357 |     900 |
| Low                   |        113 |       230 |        233 |     576 |
| Medium                |        319 |       604 |        600 |    1523 |
| Total                 |        647 |      1162 |       1190 |    2999 |

=================== Access_to_Resources vs Learning_Disabilities ===================
| Access_to_Resources   |   No |   Yes |   Total |
|:----------------------|-----:|------:|--------:|
| High                  |  799 |   101 |     900 |
| Low                   |  518 |    58 |     576 |
| Medium                | 1354 |   169 |    1523 |
| Total                 | 2671 |   328 |    2999 |

=================== Access_to_Resources vs Parental_Education_Level ===================
| Access_to_Resources   |   College |   High School |   Postgraduate |   Total |
|:----------------------|----------:|--------------:|---------------:|--------:|
| High                  |       258 |           450 |            178 |     886 |
| Low                   |       166 |           299 |            104 |     569 |
| Medium                |       461 |           731 |            303 |    1495 |
| Total                 |       885 |          1480 |            585 |    2950 |

=================== Access_to_Resources vs Distance_from_Home ===================
| Access_to_Resources   |   Far |   Moderate |   Near |   Total |
|:----------------------|------:|-----------:|-------:|--------:|
| High                  |    93 |        285 |    511 |     889 |
| Low                   |    73 |        151 |    346 |     570 |
| Medium                |   140 |        481 |    889 |    1510 |
| Total                 |   306 |        917 |   1746 |    2969 |

=================== Access_to_Resources vs Gender ===================
| Access_to_Resources   |   Female |   Male |   Total |
|:----------------------|---------:|-------:|--------:|
| High                  |      397 |    503 |     900 |
| Low                   |      243 |    332 |     575 |
| Medium                |      630 |    893 |    1523 |
| Total                 |     1270 |   1728 |    2998 |

=================== Extracurricular_Activities vs Motivation_Level ===================
| Extracurricular_Activities   |   High |   Low |   Medium |   Total |
|:-----------------------------|-------:|------:|---------:|--------:|
| No                           |    249 |   368 |      627 |    1244 |
| Yes                          |    351 |   490 |      913 |    1754 |
| Total                        |    600 |   858 |     1540 |    2998 |

=================== Extracurricular_Activities vs Internet_Access ===================
| Extracurricular_Activities   |   No |   Yes |   Total |
|:-----------------------------|-----:|------:|--------:|
| No                           |  101 |  1143 |    1244 |
| Yes                          |  140 |  1614 |    1754 |
| Total                        |  241 |  2757 |    2998 |

=================== Extracurricular_Activities vs Family_Income ===================
| Extracurricular_Activities   |   High |   Low |   Medium |   Total |
|:-----------------------------|-------:|------:|---------:|--------:|
| No                           |    233 |   527 |      483 |    1243 |
| Yes                          |    327 |   726 |      702 |    1755 |
| Total                        |    560 |  1253 |     1185 |    2998 |

=================== Extracurricular_Activities vs Teacher_Quality ===================
| Extracurricular_Activities   |   High |   Low |   Medium |   Total |
|:-----------------------------|-------:|------:|---------:|--------:|
| No                           |    373 |   130 |      727 |    1230 |
| Yes                          |    514 |   168 |     1048 |    1730 |
| Total                        |    887 |   298 |     1775 |    2960 |

=================== Extracurricular_Activities vs School_Type ===================
| Extracurricular_Activities   |   Private |   Public |   Total |
|:-----------------------------|----------:|---------:|--------:|
| No                           |       393 |      851 |    1244 |
| Yes                          |       525 |     1229 |    1754 |
| Total                        |       918 |     2080 |    2998 |

=================== Extracurricular_Activities vs Peer_Influence ===================
| Extracurricular_Activities   |   Negative |   Neutral |   Positive |   Total |
|:-----------------------------|-----------:|----------:|-----------:|--------:|
| No                           |        262 |       519 |        463 |    1244 |
| Yes                          |        384 |       644 |        727 |    1755 |
| Total                        |        646 |      1163 |       1190 |    2999 |

=================== Extracurricular_Activities vs Learning_Disabilities ===================
| Extracurricular_Activities   |   No |   Yes |   Total |
|:-----------------------------|-----:|------:|--------:|
| No                           | 1112 |   132 |    1244 |
| Yes                          | 1559 |   196 |    1755 |
| Total                        | 2671 |   328 |    2999 |

=================== Extracurricular_Activities vs Parental_Education_Level ===================
| Extracurricular_Activities   |   College |   High School |   Postgraduate |   Total |
|:-----------------------------|----------:|--------------:|---------------:|--------:|
| No                           |       345 |           623 |            256 |    1224 |
| Yes                          |       539 |           858 |            329 |    1726 |
| Total                        |       884 |          1481 |            585 |    2950 |

=================== Extracurricular_Activities vs Distance_from_Home ===================
| Extracurricular_Activities   |   Far |   Moderate |   Near |   Total |
|:-----------------------------|------:|-----------:|-------:|--------:|
| No                           |   128 |        367 |    738 |    1233 |
| Yes                          |   178 |        550 |   1008 |    1736 |
| Total                        |   306 |        917 |   1746 |    2969 |

=================== Extracurricular_Activities vs Gender ===================
| Extracurricular_Activities   |   Female |   Male |   Total |
|:-----------------------------|---------:|-------:|--------:|
| No                           |      543 |    700 |    1243 |
| Yes                          |      727 |   1028 |    1755 |
| Total                        |     1270 |   1728 |    2998 |

=================== Motivation_Level vs Internet_Access ===================
| Motivation_Level   |   No |   Yes |   Total |
|:-------------------|-----:|------:|--------:|
| High               |   47 |   553 |     600 |
| Low                |   81 |   777 |     858 |
| Medium             |  113 |  1427 |    1540 |
| Total              |  241 |  2757 |    2998 |

=================== Motivation_Level vs Family_Income ===================
| Motivation_Level   |   High |   Low |   Medium |   Total |
|:-------------------|-------:|------:|---------:|--------:|
| High               |    109 |   251 |      240 |     600 |
| Low                |    165 |   362 |      331 |     858 |
| Medium             |    285 |   640 |      615 |    1540 |
| Total              |    559 |  1253 |     1186 |    2998 |

=================== Motivation_Level vs Teacher_Quality ===================
| Motivation_Level   |   High |   Low |   Medium |   Total |
|:-------------------|-------:|------:|---------:|--------:|
| High               |    197 |    58 |      341 |     596 |
| Low                |    253 |    71 |      522 |     846 |
| Medium             |    437 |   169 |      912 |    1518 |
| Total              |    887 |   298 |     1775 |    2960 |

=================== Motivation_Level vs School_Type ===================
| Motivation_Level   |   Private |   Public |   Total |
|:-------------------|----------:|---------:|--------:|
| High               |       179 |      421 |     600 |
| Low                |       276 |      582 |     858 |
| Medium             |       464 |     1076 |    1540 |
| Total              |       919 |     2079 |    2998 |

=================== Motivation_Level vs Peer_Influence ===================
| Motivation_Level   |   Negative |   Neutral |   Positive |   Total |
|:-------------------|-----------:|----------:|-----------:|--------:|
| High               |        133 |       225 |        242 |     600 |
| Low                |        184 |       325 |        349 |     858 |
| Medium             |        329 |       613 |        599 |    1541 |
| Total              |        646 |      1163 |       1190 |    2999 |

=================== Motivation_Level vs Learning_Disabilities ===================
| Motivation_Level   |   No |   Yes |   Total |
|:-------------------|-----:|------:|--------:|
| High               |  525 |    75 |     600 |
| Low                |  774 |    84 |     858 |
| Medium             | 1372 |   169 |    1541 |
| Total              | 2671 |   328 |    2999 |

=================== Motivation_Level vs Parental_Education_Level ===================
| Motivation_Level   |   College |   High School |   Postgraduate |   Total |
|:-------------------|----------:|--------------:|---------------:|--------:|
| High               |       185 |           304 |            103 |     592 |
| Low                |       256 |           406 |            178 |     840 |
| Medium             |       444 |           770 |            304 |    1518 |
| Total              |       885 |          1480 |            585 |    2950 |

=================== Motivation_Level vs Distance_from_Home ===================
| Motivation_Level   |   Far |   Moderate |   Near |   Total |
|:-------------------|------:|-----------:|-------:|--------:|
| High               |    59 |        187 |    350 |     596 |
| Low                |    95 |        265 |    490 |     850 |
| Medium             |   152 |        465 |    906 |    1523 |
| Total              |   306 |        917 |   1746 |    2969 |

=================== Motivation_Level vs Gender ===================
| Motivation_Level   |   Female |   Male |   Total |
|:-------------------|---------:|-------:|--------:|
| High               |      253 |    346 |     599 |
| Low                |      374 |    484 |     858 |
| Medium             |      642 |    899 |    1541 |
| Total              |     1269 |   1729 |    2998 |

=================== Internet_Access vs Family_Income ===================
| Internet_Access   |   High |   Low |   Medium |   Total |
|:------------------|-------:|------:|---------:|--------:|
| No                |     46 |   102 |       93 |     241 |
| Yes               |    514 |  1150 |     1093 |    2757 |
| Total             |    560 |  1252 |     1186 |    2998 |

=================== Internet_Access vs Teacher_Quality ===================
| Internet_Access   |   High |   Low |   Medium |   Total |
|:------------------|-------:|------:|---------:|--------:|
| No                |     64 |    25 |      149 |     238 |
| Yes               |    823 |   273 |     1626 |    2722 |
| Total             |    887 |   298 |     1775 |    2960 |

=================== Internet_Access vs School_Type ===================
| Internet_Access   |   Private |   Public |   Total |
|:------------------|----------:|---------:|--------:|
| No                |        77 |      164 |     241 |
| Yes               |       842 |     1915 |    2757 |
| Total             |       919 |     2079 |    2998 |

=================== Internet_Access vs Peer_Influence ===================
| Internet_Access   |   Negative |   Neutral |   Positive |   Total |
|:------------------|-----------:|----------:|-----------:|--------:|
| No                |         41 |       101 |         99 |     241 |
| Yes               |        606 |      1062 |       1090 |    2758 |
| Total             |        647 |      1163 |       1189 |    2999 |

=================== Internet_Access vs Learning_Disabilities ===================
| Internet_Access   |   No |   Yes |   Total |
|:------------------|-----:|------:|--------:|
| No                |  217 |    24 |     241 |
| Yes               | 2454 |   304 |    2758 |
| Total             | 2671 |   328 |    2999 |

=================== Internet_Access vs Parental_Education_Level ===================
| Internet_Access   |   College |   High School |   Postgraduate |   Total |
|:------------------|----------:|--------------:|---------------:|--------:|
| No                |        86 |           109 |             43 |     238 |
| Yes               |       798 |          1372 |            542 |    2712 |
| Total             |       884 |          1481 |            585 |    2950 |

=================== Internet_Access vs Distance_from_Home ===================
| Internet_Access   |   Far |   Moderate |   Near |   Total |
|:------------------|------:|-----------:|-------:|--------:|
| No                |    21 |         70 |    149 |     240 |
| Yes               |   285 |        846 |   1598 |    2729 |
| Total             |   306 |        916 |   1747 |    2969 |

=================== Internet_Access vs Gender ===================
| Internet_Access   |   Female |   Male |   Total |
|:------------------|---------:|-------:|--------:|
| No                |      110 |    131 |     241 |
| Yes               |     1160 |   1597 |    2757 |
| Total             |     1270 |   1728 |    2998 |

=================== Family_Income vs Teacher_Quality ===================
| Family_Income   |   High |   Low |   Medium |   Total |
|:----------------|-------:|------:|---------:|--------:|
| High            |    161 |    58 |      338 |     557 |
| Low             |    367 |   129 |      738 |    1234 |
| Medium          |    359 |   111 |      699 |    1169 |
| Total           |    887 |   298 |     1775 |    2960 |

=================== Family_Income vs School_Type ===================
| Family_Income   |   Private |   Public |   Total |
|:----------------|----------:|---------:|--------:|
| High            |       163 |      397 |     560 |
| Low             |       381 |      871 |    1252 |
| Medium          |       374 |      812 |    1186 |
| Total           |       918 |     2080 |    2998 |

=================== Family_Income vs Peer_Influence ===================
| Family_Income   |   Negative |   Neutral |   Positive |   Total |
|:----------------|-----------:|----------:|-----------:|--------:|
| High            |        111 |       225 |        224 |     560 |
| Low             |        287 |       479 |        487 |    1253 |
| Medium          |        248 |       459 |        479 |    1186 |
| Total           |        646 |      1163 |       1190 |    2999 |

=================== Family_Income vs Learning_Disabilities ===================
| Family_Income   |   No |   Yes |   Total |
|:----------------|-----:|------:|--------:|
| High            |  494 |    66 |     560 |
| Low             | 1133 |   120 |    1253 |
| Medium          | 1044 |   142 |    1186 |
| Total           | 2671 |   328 |    2999 |

=================== Family_Income vs Parental_Education_Level ===================
| Family_Income   |   College |   High School |   Postgraduate |   Total |
|:----------------|----------:|--------------:|---------------:|--------:|
| High            |       162 |           278 |            112 |     552 |
| Low             |       368 |           619 |            248 |    1235 |
| Medium          |       354 |           584 |            225 |    1163 |
| Total           |       884 |          1481 |            585 |    2950 |

=================== Family_Income vs Distance_from_Home ===================
| Family_Income   |   Far |   Moderate |   Near |   Total |
|:----------------|------:|-----------:|-------:|--------:|
| High            |    52 |        183 |    317 |     552 |
| Low             |   128 |        403 |    714 |    1245 |
| Medium          |   126 |        331 |    715 |    1172 |
| Total           |   306 |        917 |   1746 |    2969 |

=================== Family_Income vs Gender ===================
| Family_Income   |   Female |   Male |   Total |
|:----------------|---------:|-------:|--------:|
| High            |      227 |    333 |     560 |
| Low             |      544 |    709 |    1253 |
| Medium          |      499 |    686 |    1185 |
| Total           |     1270 |   1728 |    2998 |

=================== Teacher_Quality vs School_Type ===================
| Teacher_Quality   |   Private |   Public |   Total |
|:------------------|----------:|---------:|--------:|
| High              |       263 |      624 |     887 |
| Low               |        92 |      206 |     298 |
| Medium            |       552 |     1223 |    1775 |
| Total             |       907 |     2053 |    2960 |

=================== Teacher_Quality vs Peer_Influence ===================
| Teacher_Quality   |   Negative |   Neutral |   Positive |   Total |
|:------------------|-----------:|----------:|-----------:|--------:|
| High              |        189 |       353 |        345 |     887 |
| Low               |         53 |       121 |        124 |     298 |
| Medium            |        399 |       672 |        705 |    1776 |
| Total             |        641 |      1146 |       1174 |    2961 |

=================== Teacher_Quality vs Learning_Disabilities ===================
| Teacher_Quality   |   No |   Yes |   Total |
|:------------------|-----:|------:|--------:|
| High              |  781 |   106 |     887 |
| Low               |  267 |    31 |     298 |
| Medium            | 1588 |   188 |    1776 |
| Total             | 2636 |   325 |    2961 |

=================== Teacher_Quality vs Parental_Education_Level ===================
| Teacher_Quality   |   College |   High School |   Postgraduate |   Total |
|:------------------|----------:|--------------:|---------------:|--------:|
| High              |       263 |           440 |            173 |     876 |
| Low               |        79 |           151 |             64 |     294 |
| Medium            |       531 |           874 |            341 |    1746 |
| Total             |       873 |          1465 |            578 |    2916 |

=================== Teacher_Quality vs Distance_from_Home ===================
| Teacher_Quality   |   Far |   Moderate |   Near |   Total |
|:------------------|------:|-----------:|-------:|--------:|
| High              |    96 |        256 |    530 |     882 |
| Low               |    25 |         87 |    184 |     296 |
| Medium            |   179 |        561 |   1014 |    1754 |
| Total             |   300 |        904 |   1728 |    2932 |

=================== Teacher_Quality vs Gender ===================
| Teacher_Quality   |   Female |   Male |   Total |
|:------------------|---------:|-------:|--------:|
| High              |      379 |    508 |     887 |
| Low               |      113 |    185 |     298 |
| Medium            |      757 |   1018 |    1775 |
| Total             |     1249 |   1711 |    2960 |

=================== School_Type vs Peer_Influence ===================
| School_Type   |   Negative |   Neutral |   Positive |   Total |
|:--------------|-----------:|----------:|-----------:|--------:|
| Private       |        183 |       357 |        379 |     919 |
| Public        |        463 |       806 |        811 |    2080 |
| Total         |        646 |      1163 |       1190 |    2999 |

=================== School_Type vs Learning_Disabilities ===================
| School_Type   |   No |   Yes |   Total |
|:--------------|-----:|------:|--------:|
| Private       |  823 |    96 |     919 |
| Public        | 1848 |   232 |    2080 |
| Total         | 2671 |   328 |    2999 |

=================== School_Type vs Parental_Education_Level ===================
| School_Type   |   College |   High School |   Postgraduate |   Total |
|:--------------|----------:|--------------:|---------------:|--------:|
| Private       |       276 |           436 |            191 |     903 |
| Public        |       608 |          1045 |            394 |    2047 |
| Total         |       884 |          1481 |            585 |    2950 |

=================== School_Type vs Distance_from_Home ===================
| School_Type   |   Far |   Moderate |   Near |   Total |
|:--------------|------:|-----------:|-------:|--------:|
| Private       |    88 |        274 |    547 |     909 |
| Public        |   218 |        642 |   1200 |    2060 |
| Total         |   306 |        916 |   1747 |    2969 |

=================== School_Type vs Gender ===================
| School_Type   |   Female |   Male |   Total |
|:--------------|---------:|-------:|--------:|
| Private       |      376 |    543 |     919 |
| Public        |      894 |   1185 |    2079 |
| Total         |     1270 |   1728 |    2998 |

=================== Peer_Influence vs Learning_Disabilities ===================
| Peer_Influence   |   No |   Yes |   Total |
|:-----------------|-----:|------:|--------:|
| Negative         |  572 |    75 |     647 |
| Neutral          | 1034 |   129 |    1163 |
| Positive         | 1066 |   124 |    1190 |
| Total            | 2672 |   328 |    3000 |

=================== Peer_Influence vs Parental_Education_Level ===================
| Peer_Influence   |   College |   High School |   Postgraduate |   Total |
|:-----------------|----------:|--------------:|---------------:|--------:|
| Negative         |       214 |           300 |            123 |     637 |
| Neutral          |       322 |           596 |            226 |    1144 |
| Positive         |       349 |           585 |            236 |    1170 |
| Total            |       885 |          1481 |            585 |    2951 |

=================== Peer_Influence vs Distance_from_Home ===================
| Peer_Influence   |   Far |   Moderate |   Near |   Total |
|:-----------------|------:|-----------:|-------:|--------:|
| Negative         |    73 |        196 |    372 |     641 |
| Neutral          |   110 |        339 |    701 |    1150 |
| Positive         |   123 |        382 |    674 |    1179 |
| Total            |   306 |        917 |   1747 |    2970 |

=================== Peer_Influence vs Gender ===================
| Peer_Influence   |   Female |   Male |   Total |
|:-----------------|---------:|-------:|--------:|
| Negative         |      284 |    363 |     647 |
| Neutral          |      495 |    668 |    1163 |
| Positive         |      491 |    698 |    1189 |
| Total            |     1270 |   1729 |    2999 |

=================== Learning_Disabilities vs Parental_Education_Level ===================
| Learning_Disabilities   |   College |   High School |   Postgraduate |   Total |
|:------------------------|----------:|--------------:|---------------:|--------:|
| No                      |       793 |          1310 |            528 |    2631 |
| Yes                     |        92 |           171 |             57 |     320 |
| Total                   |       885 |          1481 |            585 |    2951 |

=================== Learning_Disabilities vs Distance_from_Home ===================
| Learning_Disabilities   |   Far |   Moderate |   Near |   Total |
|:------------------------|------:|-----------:|-------:|--------:|
| No                      |   273 |        825 |   1549 |    2647 |
| Yes                     |    33 |         92 |    198 |     323 |
| Total                   |   306 |        917 |   1747 |    2970 |

=================== Learning_Disabilities vs Gender ===================
| Learning_Disabilities   |   Female |   Male |   Total |
|:------------------------|---------:|-------:|--------:|
| No                      |     1126 |   1546 |    2672 |
| Yes                     |      144 |    183 |     327 |
| Total                   |     1270 |   1729 |    2999 |

=================== Parental_Education_Level vs Distance_from_Home ===================
| Parental_Education_Level   |   Far |   Moderate |   Near |   Total |
|:---------------------------|------:|-----------:|-------:|--------:|
| College                    |    89 |        276 |    511 |     876 |
| High School                |   148 |        443 |    876 |    1467 |
| Postgraduate               |    64 |        178 |    336 |     578 |
| Total                      |   301 |        897 |   1723 |    2921 |

=================== Parental_Education_Level vs Gender ===================
| Parental_Education_Level   |   Female |   Male |   Total |
|:---------------------------|---------:|-------:|--------:|
| College                    |      375 |    510 |     885 |
| High School                |      641 |    839 |    1480 |
| Postgraduate               |      232 |    353 |     585 |
| Total                      |     1248 |   1702 |    2950 |

=================== Distance_from_Home vs Gender ===================
| Distance_from_Home   |   Female |   Male |   Total |
|:---------------------|---------:|-------:|--------:|
| Far                  |      136 |    170 |     306 |
| Moderate             |      381 |    535 |     916 |
| Near                 |      739 |   1008 |    1747 |
| Total                |     1256 |   1713 |    2969 |